File: compiler/bench/textwrap.c
// expected: 5
// Text wrap benchmark - wrap text at width, count lines, 50000 times

#include <stdio.h>
#include <string.h>

char text[128];

void init_text(void) {
    strcpy(text, "The quick brown fox jumps over the lazy dog near the river bank");
}

int wrap_count(const char *s, int len, int width) {
    int lines = 1;
    int col = 0;
    int last_space = 0;
    for (int i = 0; i < len; i++) {
        if (s[i] == ' ') {
            last_space = i;
        }
        if (col >= width) {
            lines++;
            col = 0;
        } else {
            col++;
        }
    }
    return lines;
}

int main(void) {
    init_text();
    int result = 0;
    for (int i = 0; i < 50000; i++) {
        result = wrap_count(text, 64, 12);
    }
    printf("%d\n", result);
    return 0;
}

File: compiler/bench/textwrap.fs
\ expected: 5
\ Text wrap benchmark - wrap text at width, count lines, 50000 times

create text 128 allot

: init-text ( -- )
  s" The quick brown fox jumps over the lazy dog near the river bank" drop text 64 cmove ;

: wrap-count ( addr len width -- lines )
  1 >r
  0 0 do
    over i + c@ 32 = if
      drop i
    then
    dup 2 pick >= if
      r> 1+ >r
      drop 0
    else
      1+
    then
  loop
  2drop r> ;

: main ( -- )
  init-text
  0
  50000 0 do
    drop
    text 64 12 wrap-count
  loop
  . cr ;
